azure-proto-core-test/IdentityTests.cs

```diff
@@ -0,0 +1,107 @@
+        [TestCase("/subscriptions/6b085460-5f21-477e-ba44-1035046e9101", null, true)]
+        public void CompareToOne(string resourceId1, string resourceId2, bool isNull)
+        {
+            Identity identity1 = new Identity();
+            identity1.ResourceId = resourceId1;
+            Identity identity2 = new Identity();
+            identity2.ResourceId = resourceId2;
+            Assert.AreEqual(1, identity1.CompareTo(isNull?null: identity2));           
+        }
+
+        [TestCase("/subscriptions/6b085460-5f21-477e-ba44-1035046e9101", "/subscriptions/6b085460-5f21-477e-ba44-1035046e9101/resourceGroups/nbhatia_test")]
+        [TestCase("/subscriptions/6b085460-5f21-477e-ba44-1035046e9101", "/subscriptions/6b085460-5f21-477e-ba44-1035046e9101/resourceGroups/nbhatia_test/providers/Microsoft.Web/sites/autoreport")]
+        public void CompareToMinusOne(string resourceId1, string resourceId2)
+        {
+            Identity identity1 = new Identity();
+            identity1.ResourceId = resourceId1;
+            Identity identity2 = new Identity();
+            identity2.ResourceId = resourceId2;
+            Assert.AreEqual(-1, identity1.CompareTo(identity2));
+        }
+
+        [TestCase]
+        public void EqualsMethodTrue()
+        {
+            Identity identity1 = new Identity();
+            Identity identity2 = new Identity();
+            
+            identity1.ResourceId = "/subscriptions/6b085460-5f21-477e-ba44-1035046e9101/resourceGroups/nbhatia_test/providers/Microsoft.Web/sites/autoreport";
+            identity2.ResourceId = "/subscriptions/6b085460-5f21-477e-ba44-1035046e9101/resourceGroups/nbhatia_test/providers/Microsoft.Web/sites/autoreport";
+
+            Guid tenantId = Guid.NewGuid();
+            Guid clientId = Guid.NewGuid();
+            Guid principalId = Guid.NewGuid();
+            identity1.ClientId = clientId;
+            identity1.TenantId = tenantId;
+            identity1.PrincipalId = principalId;
+            identity2.ClientId = clientId;
+            identity2.TenantId = tenantId;
+            identity2.PrincipalId = principalId;
+            identity1.Kind = new IdentityKind("test");
+            identity2.Kind = new IdentityKind("test");
+
+            Assert.AreEqual(identity1.Equals(identity2), true);
+
+            Identity identity3 = new Identity();
+            Identity identity4 = new Identity();
+            Assert.AreEqual(identity1.Equals(identity2), true);
+        }
+
+        [TestCase]
+        public void EqualsMethodFalse()
+        {
+            Identity identity1 = new Identity();
+            Identity identity2 = new Identity();
+
+            identity1.ResourceId = "/subscriptions/6b085460-5f21-477e-ba44-1035046e9101/resourceGroups/nbhatia_test/providers/Microsoft.Web/sites/autoreport";
```

Comment:
In general, it makes sense to think about the equivalency classes in inputs like equal and non-equal identities, for example
(0) identities that are null or empty (if empty is possible)
(1) identities that consist of exactly the same characters
(2) identities that are equivalent but not identical (identities that vary by capitalization)

and make sure we have tests for these, irrespective of code coverage - this way we have regression protection for important scenarios (like case-insensitive comparison), regardless of the implementation details